src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs

```diff
@@ -68,16 +65,15 @@ internal static SafeFileHandle Open(string path, Interop.Sys.OpenFlags flags, in
 
         /// <summary>Opens a SafeFileHandle for a file descriptor created by a provided delegate.</summary>
         /// <param name="fdFunc">
-        /// The function that creates the file descriptor. Returns the file descriptor on success, or -1 on error,
-        /// with Marshal.GetLastWin32Error() set to the error code.
+        /// The function that creates the file descriptor. Returns the file descriptor on success, or an invalid
+        /// file descriptor on error with Marshal.GetLastWin32Error() set to the error code.
         /// </param>
         /// <returns>The created SafeFileHandle.</returns>
-        internal static SafeFileHandle Open(Func<IntPtr> fdFunc)
+        internal static SafeFileHandle Open(Func<SafeFileHandle> fdFunc)
         {
-            var handle = new SafeFileHandle(ownsHandle: true);
-            IntPtr fd;
-            while (Interop.CheckIo(fd = fdFunc())) ;
-            handle.SetHandle(fd);
+            SafeFileHandle handle = fdFunc();
+            Interop.CheckIo(handle);
```

Comment:
You might consider having CheckIo return its first argument, now that it doesn't need to return a bool.  Then cases like this become:
```C#
SafeFileHandle handle = Interop.CheckIo(fdFunc());
```